docs/source/general/freeradius.rst

```diff
@@ -208,6 +209,76 @@ An example using MySQL is:
                              AND valid_until >= CURDATE() \
                              ORDER BY id"
 
+Using Radius Checks for Authorization Information
+-------------------------------------------------
+
+Traditionally, when using an SQL backend with Freeradius, user authorization information such as User-Name and
+`"known good" <https://freeradius.org/radiusd/man/rlm_pap.html>`_ password are stored using the *radcheck*
+table provided by Freeradius' default SQL schema.  Django-Freeradius utilizes Freeradius'
+`rlm_rest <https://networkradius.com/doc/current/raddb/mods-available/rest.html>`_ module in order to
+take advantage of the built in user management and authentication capabilities of Django.
+(See :ref:`configure-rest-module` and `User authentication in Django <https://docs.djangoproject.com/en/dev/topics/auth/>`_)
+
+For existing Freeradius deployments or in cases where it is preferred to utilize Freeradius' *radcheck* table for
+storing user credentials it is possible to utilize `rlm_sql <https://wiki.freeradius.org/modules/Rlm_sql>`_
+instead of `rlm_rest <https://networkradius.com/doc/current/raddb/mods-available/rest.html>`_ for authorization.
+
+.. note::
+    Bypassing the Django-Freeradius' REST API for authorization means you will have to manually create
+    Radius Check 'password' entries for each user you want to authenticate with Freeradius.
+
+Password hashing
+^^^^^^^^^^^^^^^^
+
+By default Django will use `PBKDF2 <https://en.wikipedia.org/wiki/PBKDF2>`_ to store all passwords in the database.
+(See `Password management in Django <https://docs.djangoproject.com/en/dev/topics/auth/passwords/)>`_).
+The default password hashing and storage algorithms in Django are not compatible with those used by Freeradius.
+Therefore, a default set of Freeradius compatible password storage methods have been provided for deployments that make use
+of Radius Checks for user credentials.
+
+* Cleartext-Password
+* NT-Password
+* LM-Password
+* MD5-Password
+* SMD5-Password
+* SHA-Password
+* SSHA-Password
+* Crypt-Password
+
+.. note::
+    Only the Crypt-Password hashing attribute is recommended for new entries as it makes
+    use of the sha512_crypt feature supported by most Unix/Linux operating systems.
+    (See `passlib.hash <https://passlib.readthedocs.io/en/stable/lib/passlib.hash.html#active-unix-hashes>`_)
+    The other password hashing algorithms have been provided for backward compatibility.
+
+Configuration
+^^^^^^^^^^^^^
+
+To configure support for accessing user credentials with Radius Checks edit the ``authorize`` section
+of the ``/etc/freeradius/3.0/sites-available/default`` configuration file as follows
+
+.. code-block:: ini
+
+    # /etc/freeradius/3.0/sites-available/default
+
+    authorize {
+       #rest
```

Comment:
one question: can we implement a way to use both? If yes do you have any idea how?
@2stacks @peppelinux

Reply:
I believe so.  I mentioned somewhere earlier that the reason it currently fails is that the Rest API returns Auth-Type := Accept and/or Reject.  Doing this prevents Freeradius from being able to perform Authentication via PAP which is how credentials in radcheck are supposed to be handled.  Below is some of the inline documentation taken from the 'default' virtual server that Freeradius uses.

#  In general, you SHOULD NOT set the Auth-Type attribute.  The server
#  will figure it out on its own, and will do the right thing.  The
#  most common side effect of erroneously setting the Auth-Type
#  attribute is that one authentication method will work, but the
#  others will not.

I don't know that I have the skill to change how the Rest API is currently implemented but I believe it's possible to support authentication against auth_user and radcheck tables simultaneously.  I'll take a look and see if it's something I can do.

Reply:
@2stacks thanks for explaining, now my understanding has increased.
Try hacking this line: https://github.com/openwisp/django-freeradius/blob/master/django_freeradius/api/views.py#L48
That's the line that explicitly sets the `Auth-Type := Reject`

Reply:
Thanks, I'll start looking there.  I'm pretty sure we will still have to set Auth-Type := Accept since the django hashes wont work with PAP.  If a user doesn't exist in the auth_user table or their credentials can't be verified I think instead of returning Auth-Type := Reject we just let it continue normal freeradius evaluation.  Not sure what not setting an Auth-Type will do but that's what I'm going to find out :)